### PR TITLE
redirect from csp to csp-2017 and from csd to csd-2017

### PR DESCRIPTION
**Pull request by davidsbailey**
### Description

* start serving csp at csp-2017 and csd at csd-2017, without actually renaming the underlying courses just yet. do this in a way that will still work when the underlying course is renamed.
* start redirecting csp --> csp-2017 and csd --> csd-2017

### Background
* [dev doc](https://docs.google.com/document/d/1Os6zB01TophoitiOdE6quf5SH1H0TW4gZFdv4EknMrg/edit#) outlining the goal and next steps
* [timeline](https://docs.google.com/document/d/1c7clVeMH794x3CVM5gRPKdUD6SwoPx_I5qtmcc33dnI/edit#heading=h.5vx4d87coar1) (what needs to be done by June 1)
* [course terminology gsheet](https://docs.google.com/spreadsheets/d/1naiSH_RfWw-RijhbWKy4ZwNOctZBtd2Cm7RegL0bPdY/edit#gid=0)
* the [spec](https://docs.google.com/document/d/14RnmNy9QngJB8Y6mMd7IDiYrFs-_fNgr_zXklJ-MsUU/edit#) , which you can probably get away without reading

### CODE REVIEW

**Comment by davidsbailey** [comment]
Hi Brad and Ram, this is part pull request and part design review for the [dev doc](https://docs.google.com/document/d/1Os6zB01TophoitiOdE6quf5SH1H0TW4gZFdv4EknMrg/edit). I haven't written any tests yet, because I want to get your feedback on the approach first. I'm taking some shortcuts here, due to the upcoming June 1 deadline.

File: dashboard/app/models/course.rb [hunk 1]
```diff
@@ -387,6 +387,18 @@ def self.course_cache
   end
 
   def self.get_without_cache(id_or_name)
+    # When the caller requests csp-2017, make sure we serve the CSP 2017 course,
+    # regardless of whether it has been renamed from csp to csp-2017 yet.
+    if id_or_name == 'csp-2017'
+      return Course.where(name: ['csp', 'csp-2017']).first
```
Comment (islemaster) [inline dashboard/app/models/course.rb]: [Subset conditions](http://guides.rubyonrails.org/active_record_querying.html#subset-conditions) - equivalent to `WHERE (name IN ('csp', 'csp-2017'))` - TIL!

**Comment by davidsbailey** [comment]
Thanks Brad! @balderdash , I've also sent out https://github.com/code-dot-org/code-dot-org/pull/22519 so you can see what step 2 would look like when evaluating the overall plan.

**Comment by davidsbailey** [comment]
Two quick updates:
* make it possible to rename csp-2017 --> csp during seed, so that the right thing will happen if the next PR which renames csp --> csp-2017 is reverted.
* defensively make Course.get_without_cache('csp') look for 'csp-2017' as well as 'csp', so that if we miss some call to `Course.get_without_cache('csp')` during the next PR, it won't immediately break anything.

**Comment by balderdash** [comment]
This should also be updated to accommodate the versioned courses:
https://github.com/code-dot-org/code-dot-org/blob/f22d91304cbce9731dd487b1fa742b1079fddabe/apps/src/templates/teacherDashboard/sections.js#L121-L125

Looks like it's currently broken for sections assigned to the `*-2018` versions of csp and csd.

**Comment by davidsbailey** [comment]
I've verified locally that the right thing will show up at /courses/csp-2017 when csp is renamed to csp-2017, and then again after it is renamed back to csp (this is what would happen if https://github.com/code-dot-org/code-dot-org/pull/22519 was merged and then reverted).

I'm going to address the issues with teacherDashboard/sections.js in the next PR, because this PR doesn't make things worse and I want this PR to be as easy to revert as possible if something goes wrong.

**Pull request merged**